Automattic/wp-calypso
# Cache latest JITMs in state to reduce SiteNotice and JITM component rerenders

This PR is part of [a wider effort to optimize the `SiteNotices` and `JITM` components to avoid the "jumpiness" of the sidebar navigation experience](https://github.com/Automattic/wp-calypso/issues/46091) both on the current production version of Calypso sidebar menu _and_ the upcoming "Unified Navigation" version of the menu.

Currently `SiteNotice` will render often due to x2 props which change frequently:

* `hasJITM`
* `messagePath`

Each time these props change the entire component renders which causes the entire "Notices/Nudges" area to be removed from the DOM and then reinserted a few seconds later. This causes the entire nav to "jump".

This PR seeks to optimize the `hasJITM` case. The [`messagePath` issue is addressed in another PR](https://github.com/Automattic/wp-calypso/pull/46676).

⚠️ **Update**: it looks like we may not need to optimise the `messagePath` part of the problem. This PR works and resolves the jumping of the sidebar menu _without_ having any optimisations for `messagePath`.

Essentially the result of calling `getTopJITM` changes frequently between `null` and a populated JITM object. The result of this is that both `SiteNotice` and `JITM` component re-render often causing the sidebar menu to "jump" as the component is added/removed from the DOM. 

Why this happens best explained by @mreishus [here](https://github.com/Automattic/wp-calypso/pull/46448#issuecomment-713087683) so I will quote verbatim from him below in `How do JITMs work`.

### How do JITMs work (by @mreishus)

JITM stands for "Just In Time Messages" and is a flexible system that communicates with the backend to show messages like "Free Domain with a plan [Upgrade]".  It works on a per site,per page basis.

Let's say I visit the "posts" section in calypso.  It fires off this event:
```
{
  type: 'JITM_FETCH',
  siteId: 181129030,
  messagePath: 'calypso:posts:sidebar_notice',
  locale: 'en',
  meta: {
    dataLayer: {
      doBypass: true
    }
  }
}
```
Which sends an API request, and ends up filling `state.jitm.sitePathJITM["calypso:posts:sidebar_notice181129030"] = [ ...arrayofmessageshere... ]`

The sidebar looks in that part of the state and displays the upsell message.  Now, if I click on the "pages" section, the sidebar starts looking for `state.jitm.sitePathJITM["calypso:pages:sidebar_notice181129030"]`, which is empty, so the upsell disappears.  It sends another event which does another API request:
```
{
  type: 'JITM_FETCH',
  siteId: 181129030,
  messagePath: 'calypso:pages:sidebar_notice',
  locale: 'en',
  meta: {
    dataLayer: {
      doBypass: true
    }
  }
}
```
This takes a while to return, but when it does, then the `state.jitm.sitePathJITM["calypso:pages:sidebar_notice181129030"]` part of state is filled in and we see the same upsell.  So it's a bit silly and jarring - Upon navigation, we saw the upsell message disappear and then after a few seconds, the exact same message is redisplayed.

## How does this PR fix this?

@getdave here again! 😄 Picking up from where @mreishus left off above...

This PR fixes the issue by maintaining a cached copy of the last _valid_  (ie: not `null` or empty) JITM message that was received for a given `siteId`. This cache is ignorant of `messagePath` - it is concerned only with caching messages by `siteId`. 

When the `getTopJITM` selector is called, it checks to see whether the state has been populated with JITMs for the given `messagePath`. 

If the state is `null` at the given `messagePath` we know the JITMs have yet to be populated from the network for the given path and so we check the JITM cache to see if we can fallback to the most recent JITM that was available for the given `siteId`.

If a cached JITM is found, we return that JITM object value. It is this action that avoids the `getTopJITM` selector from returning a `null` value. This stops the `SiteNotice` and `JITM` components from re-rendering unecessarily.

If once the network request for JITMs at a given `messagePath` fails the state will be populated with an empty array to better signify a "populated but empty" state. At this point the `getTopJITM` will return `null` and the component will re-render ensuring that "stale" cached JITM values are not allowed to persist in the UI.

## Changes proposed in this Pull Request

* Adds JITMs cache to state for a given `siteId` (`sitePathJITMCache`).
* Update `getJITM` selector to return `null` when given `messagePath` is yet to be populated ("empty").
* Update `getTopJITM` selector to return value from  JITM cache if given `messagePath` is truly "empty" (ie: `null`). Ignore cache if it is also empty.
* Ensure clearing of JITMs cache from `clearJITM` action.
* Ensure JITM cache is cleared when network returns no JITMs (ie: empty array).

## Screenshots

![Screen Capture on 2020-10-22 at 14-14-18](https://user-images.githubusercontent.com/444434/96884176-1e7c6800-1479-11eb-903a-5d4136353191.gif)


## Testing instructions

* Checkout PR.
*  `yarn && yarn start`. Wait for Calypso to boot.
* Goto `http://calypso.localhost:3000/home/` and select your test site.
* Add `?flags=nav-unification` to the end of the URL. Hit `Enter` to reload the page.
* Click around the sidebar menu. Notice how the Nudges/Notices doesn't disappear/re-appear on navigation to a new section of Calypso.

### Devtools test

* Open Devtools Console panel
* Inspect `state.jitm` - see two properties:
  - `sitePathJITM` - should be populated with an object keyed by `messagePath + siteId` (eg: `calypso:pages:sidebar_notice174746428`...etc) containing JITMs for your given `messagePath` + `siteId`.
  - `sitePathJITMCache` - should be populated with an object keyed by `siteId` (eg: `174746428`) containing a single JITM 
* Now use the devtools to manually insert/clear JITMs:
  - `window._jitm.insert('calypso:home:sidebar_notice', [ // JITM objects here - see below for example ])`
  - `window._jitm.clear('calypso:home:sidebar_notice')`

Here's a JITM object I prepared earlier:

```js
{
  "message": "Dave loves JITMs",
  "description": "Aren't JITMS great?",
  "classes": "",
  "icon": "",
  "featureClass": "upsell_nudge",
  "CTA": {
    "message": "Upgrade",
    "link": "/domains/add/wordpress.com/?ref=upgrade"
  },
  "tracks": {
    "display": {
      "name": "calypso_upgrade_nudge_impression",
      "props": {
        "cta_name": "free-to-paid-sidebar"
      }
    },
    "click": {
      "name": "calypso_upgrade_nudge_cta_click",
      "props": {
        "cta_name": "free-to-paid-sidebar"
      }
    }
  },
  "template": "sidebar-banner",
  "id": "free_to_paid_plan",
  "isDismissible": true,
  "lastUpdated": 1603372456597
}
```

* You should be able to add/remove JITMs at your given `messagePath` and see the UI update accordingly.
* ⚠️ Please be careful to only insert/clear JITMs are your current `messagePath`. Remember the middle portion of the string changes on a per route basis (eg: `calypso:home:sidebar_notice`, `calypso:pages:sidebar_notice`, `calypso:posts:sidebar_notice`). If you use the wrong `messagePath` the UI will not update!

Fixes https://github.com/Automattic/wp-calypso/issues/46091
